File: 프로그래머스/lv3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.java
import java.util.*;

class Solution {
    int[] answer;
    HashMap<String, String> parent = new HashMap<>();
    HashMap<String, Integer> idx = new HashMap<>();
    
    public void dfs(String seller, int amount){
        if (seller.equals("-"))
            return;
        
        int a = amount / 10;
        answer[idx.get(seller)] += amount - a;
        
        if (a >= 1)
            dfs(parent.get(seller), a);        
    }
    
    public int[] solution(String[] enroll, String[] referral, String[] seller, int[] amount) {
        int n = enroll.length;
        answer = new int[n];
        
        for(int i=0; i<n; i++){
            parent.put(enroll[i], referral[i]);
            idx.put(enroll[i], i);
        }
        
        for(int i=0; i<seller.length; i++)
            dfs(seller[i], amount[i]*100);
        
        return answer;
    }
}

File: 프로그래머스/lv3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.py
def bfs(seller, amount, tree, idx):
    global answer
    if seller == '-':
        return
    
    a = int(amount*0.1)
    answer[idx[seller]] += amount - a
    
    if a >= 1: # 1원 미만인 경우에는 분배 X
        bfs(tree[seller], a, tree, idx) # 10% 배분
    
def solution(enroll, referral, seller, amount):
    global answer
    answer = [0]*len(enroll)
    tree = {}
    idx = {}

    for i in range(len(enroll)):
        tree[enroll[i]] = referral[i]
        idx[enroll[i]] = i
        
    for i in range(len(seller)):
        s = seller[i]
        a = amount[i]*100
        bfs(s, a, tree, idx)
            
    return answer